machine/v1beta1/types_alibabaprovider.go

```diff
@@ -0,0 +1,320 @@
+	// You can call the DescribeVSwitches operation to query the created vSwitches.
+	VSwitch AlibabaResourceReference `json:"vSwitch,omitempty"`
+
+	// RAMRoleName is the name of the instance Resource Access Management (RAM) role. This allows the instance to perform API calls as this specified RAM role.
+	// +optional
+	RAMRoleName string `json:"ramRoleName,omitempty"`
+
+	// ResourceGroupID is the unique ID of the resource group to which to assign the instance.
+	// +optional
+	ResourceGroupID string `json:"resourceGroupId,omitempty"`
+
+	//Specifies whether to create the instance on a dedicated host. Valid values:
+	//
+	//default: creates the instance on a non-dedicated host.
+	//host: creates the instance on a dedicated host. If you do not specify the DedicatedHostID parameter, Alibaba Cloud automatically selects a dedicated host for the instance.
+	//Default value: default.
+	// +optional
+	Tenancy InstanceTenancy `json:"tenancy,omitempty"`
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	// +optional
+	UserDataSecret *corev1.LocalObjectReference `json:"userDataSecret,omitempty"`
+
+	// CredentialsSecret is a reference to the secret with alibabacloud credentials. Otherwise, defaults to permissions
+	// provided by attached RAM role where the actuator is running.
+	// +optional
+	CredentialsSecret *corev1.LocalObjectReference `json:"credentialsSecret,omitempty"`
+
+	// Tags are the set of metadata to add to an instance.
+	// +optional
+	Tags []Tag `json:"tag,omitempty"`
+}
+
+// ResourceTagReference is a reference to a specific AlibabaCloud resource by ID, or tags.
+// Only one of ID or Tags may be specified. Specifying more than one will result in
+// a validation error.
+type AlibabaResourceReference struct {
+	// ID of resource
+	// +optional
+	ID string `json:"id,omitempty"`
+
+	// Tags is a set of metadata based upon ECS object tags used to identify a resource
+	// +optional
+	Tags []Tag `json:"tags,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlibabaCloudMachineProviderConfigList contains a list of AlibabaCloudMachineProviderConfig
+// Compatibility level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=2
+type AlibabaCloudMachineProviderConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []AlibabaCloudMachineProviderConfig `json:"items"`
+}
+
+// AlibabaCloudMachineProviderStatus is the Schema for the alibabacloudmachineproviderconfig API
+// Compatibility level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=2
+//+k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type AlibabaCloudMachineProviderStatus struct {
```

Comment:
Why is status named instead of being part of another resource?

Reply:
I don't have a good explanation as to why the original author made this decision. I would go out on a limb and suggest they probably copied the structure of the other providers: 
- https://github.com/openshift/api/blob/master/machine/v1beta1/types_awsprovider.go#L242
- https://github.com/openshift/api/blob/master/machine/v1beta1/types_vsphereprovider.go#L141
- https://github.com/openshift/api/blob/master/machine/v1beta1/types_azureprovider.go#L108
- https://github.com/openshift/api/blob/master/machine/v1beta1/types_gcpprovider.go#L189

Reply:
This is an artifact if the design of the Machine API.

Because both the spec and status in this PR are only ever used in raw extensions (under machine.spec.providerSpec and machine.status.providerStatus), they are fully detached resources.

There has been convention from the past to make sure that these are embedded resources, and as such, require typemeta and objectmeta.

The type meta fields tend to be used to allow us to check the expected API version and decode correctly in objects that wouldn't necessarily know which platform to expect (eg our webhooks)